### PR TITLE
[Job] [Dashboard] Add Job Submission data to cluster snapshot

### LINKED ISSUE #22290: Add Jobs metadata to /api/snapshot

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
The existing Job info in the cluster snapshot uses the old definition of Job, which is a single Ray driver (a single `ray.init()` connection).  

In the new Job Submission protocol, a Job just specifies an entrypoint which can be any shell command.  As such a Job can have zero or multiple Ray drivers.  This means we should add a new snapshot entry corresponding to new jobs.  We'll leave the old snapshot in place for legacy jobs.

- Also fixes `get_all_jobs` by using the appropriate KV namespace, and stripping the job key KV prefix from the job ID.  It wasn't working before.

- This PR also unifies the datatype used by the GET jobs/ endpoint to be the same as the one used by the new jobs cluster snapshot.  For backwards compatibility, the `status` and `message` fields are preserved.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@tchordia Could you please give a preliminary review on the format?  The primary changes would be 

- remove `config`, since its subfields `metadata` and `namespace` are scoped to a driver, not a Job.
- promote `runtime_env` to the top level, since it's one of the main fields associated with a Job

Also, is it correct to just add a new snapshot here (`job_submission`) along the existing one (`jobs`)?

**Comment by edoakes** [comment]
@architkulkarni a job can also have a namespace & metadata, so we should probably also include those

**Comment by edoakes** [comment]
@architkulkarni it's the same as existing namespaces, though I'm now realizing we don't actually support it in the jobs REST API at the moment so no need to worry about it (maybe still design for it but mark as "future")

File: dashboard/modules/snapshot/snapshot_head.py
Comment (tchordia) [inline dashboard/modules/snapshot/snapshot_head.py:145]: can we call this job_submission_id?

**Comment by tchordia** [comment]
we definitely want to include the "metadata" field, otherwise LGTM.

File: dashboard/modules/job/common.py
Comment (edoakes) [reply dashboard/modules/job/common.py:67]: looks good. if we do this, please rename `JobStatusStorageClient` accordingly

**Review by architkulkarni — COMMENTED** [review]
@tchordia @edoakes This is passing tests and ready for review!

File: dashboard/modules/snapshot/snapshot_schema.json
Comment (architkulkarni) [inline dashboard/modules/snapshot/snapshot_schema.json:74]: Does this schema make sense (in particular, the choice of which fields are nullable?)

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (architkulkarni) [inline dashboard/modules/snapshot/tests/test_job_submission.py:82]: This is bad, we don't want to convert arbitrary strings in the job data to camelcase.  It turns out we've had this problem on the existing jobs snapshot the whole time:

```
'03000000': {'config': {'metadata': {'jobName': 'raysubmit_RcjHs96ykN3SQDB1',
                                                                     'jobSubmissionId': 'raysubmit_RcjHs96ykN3SQDB1',
                                                                     'rayTest789': '789'},
                                                        'namespace': 'fa5d38e3-a819-488d-921e-a83b62a5d873',
                                                        'runtimeEnv': {'envVars': {'RAYTest456': '456'}}},
```
The env vars are incorrect.  (The camelcase algorithm seems nontrivial; how does `RAY_TEST_456` get converted to `RAYTest456`??)

Comment (architkulkarni) [reply dashboard/modules/snapshot/tests/test_job_submission.py:82]: Previous discussion on disabling the camel casing: https://github.com/ray-project/ray/pull/17977 

It seems like everyone is okay with it, but we're worried about breaking backwards compatibility.

File: dashboard/modules/snapshot/snapshot_head.py
Comment (architkulkarni) [inline dashboard/modules/snapshot/snapshot_head.py:148]: In the old job snapshot, the metadata contains the job ID:

```
'metadata': {
    'jobName': 'raysubmit_RcjHs96ykN3SQDB1',
    'jobSubmissionId': 'raysubmit_RcjHs96ykN3SQDB1',
    'rayTest789': '789'
},
```

In the new job snapshot, `metadata` just consists of the `metadata` kwarg passed in `submit_job`:

```
'metadata': {'rayTest789': '789'}
```

Is this what we want?

Comment (edoakes) [reply dashboard/modules/snapshot/snapshot_head.py:148]: yes that looks right to me! cc @tchordia

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (architkulkarni) [reply dashboard/modules/snapshot/tests/test_job_submission.py:82]: I'd propose we don't block this PR on this issue, and instead make a separate issue to disable camelcasing in the snapshot altogether.

File: dashboard/modules/job/common.py
Comment (edoakes) [inline dashboard/modules/job/common.py:118]: huh, i've never seen `replace` before. good to know...

File: dashboard/modules/snapshot/tests/test_job_submission.py
Comment (edoakes) [reply dashboard/modules/snapshot/tests/test_job_submission.py:82]: @architkulkarni can you make sure to follow up on this? it's sure to come back to bite us...

**Review by edoakes — APPROVED** [review]
I'm wondering if it'd make sense to flatten the `JobStatusInfo` into an independent `status` and `status_message` field to match the api snapshot (and the future list_jobs API) -- did you consider that? Would it make the code better/worse? Otherwise LGTM, let's merge and follow up.

**Comment by edoakes** [comment]
@wuisawesome stamp pls?

**Review by wuisawesome — APPROVED** [review]

**Comment by architkulkarni** [comment]
> I'm wondering if it'd make sense to flatten the `JobStatusInfo` into an independent `status` and `status_message` field to match the api snapshot (and the future list_jobs API) -- did you consider that? Would it make the code better/worse? 

I don't think it's much better that way, don't have a strong preference though.  JobStatusInfo has the `__post_init__` logic to populate the message field which adds a little bit of complexity, and it's kind of nice to hide that from the JobData.

It would be good to get a review on the schema from @tchordia before merging.

File: dashboard/modules/job/common.py
Comment (architkulkarni) [reply dashboard/modules/job/common.py:118]: Yup, it also calls __post_init__ after the replacing is done.

File: dashboard/modules/snapshot/snapshot_schema.json
Comment (tchordia) [reply dashboard/modules/snapshot/snapshot_schema.json:74]: 1. do these jobs have names at all? i don't think they do, but just double checking
2. can we add an error type field here (specifically for when the status is errored?) it would be good to surface if, for example, the failure was a runtime env failure. Specifically, useful error types could be:
a. runtime env setup
b. internal system failure
c. user script error
3. what is namespace? how is it populated? i thought it's possible to have multiple sub drivers that point to different namespaces in the same ray job submission.

other than that, lgtm!

Comment (architkulkarni) [reply dashboard/modules/snapshot/snapshot_schema.json:74]: 1. I don't think so, just the Job ID.  There's no way to specify a name in the Jobs API
2. Sure, let me look into this.  You're saying to add a field to the schema in this PR, and then figure out the error plumbing in a future PR?  It seems out of the scope of this PR

3. [Namespace discussion](https://github.com/ray-project/ray/pull/22225#issuecomment-1033271264) it doesn't mean anything right now, but we'll leave it here for the future.  It'll probably be something like you set a namespace in the job submission, and then each driver reads an env var and sets its namespace to that namespace, but you can override it per driver by specifying a different namespace.

Comment (edoakes) [reply dashboard/modules/snapshot/snapshot_schema.json:74]: 1. No names ATM.
2. Good idea. Should this be an "error type" or just different top-level statuses though?
3. It's not very well-defined; maybe we should just drop namespace for now and we can add it later if it makes sense.

**Comment by edoakes** [comment]
@architkulkarni let's make sure the schema for this matches https://github.com/ray-project/ray/issues/22412 exactly BTW

File: dashboard/modules/snapshot/snapshot_head.py
Comment (edoakes) [inline dashboard/modules/snapshot/snapshot_head.py:144]: to match the get & list jobs APIs, I think this should be called `message`. Can we use the same datatclass for both to ensure consistency?

File: dashboard/modules/snapshot/snapshot_schema.json
Comment (architkulkarni) [reply dashboard/modules/snapshot/snapshot_schema.json:74]: For 2, I think I'll add error_type as a top-level field rather than increase the number of top-level statuses.  Right now with the SDK the user can do a check like `if status == FAILED`, but this type of check becomes more complicated if there are many top-level failure types.

For 3, I'll drop namespace for now.

File: dashboard/modules/snapshot/snapshot_head.py
Comment (architkulkarni) [reply dashboard/modules/snapshot/snapshot_head.py:144]: I slightly preter status_message because it's more specific; I'll update the get and list job API to have `status_message` as well.

Comment (edoakes) [reply dashboard/modules/snapshot/snapshot_head.py:144]: One consideration here is the `GET /api/jobs/status` currently returns a `message` field (I think), so it'd be a breaking change to rename it. That is ok because the API is still alpha until 1.12 but it's still preferable to avoid the breakage.

**Comment by architkulkarni** [comment]
- I flattened the 

> > I'm wondering if it'd make sense to flatten the `JobStatusInfo` into an independent `status` and `status_message` field to match the api snapshot (and the future list_jobs API) -- did you consider that? Would it make the code better/worse?
> 
> I don't think it's much better that way, don't have a strong preference though. JobStatusInfo has the `__post_init__` logic to populate the message field which adds a little bit of complexity, and it's kind of nice to hide that from the JobData.
> 
 
I ended up flattening this because we're unifying it with the REST API endpoint which has `status` and `message` as top level fields, and we need to preserve backwards compatibility.

The datatype used by the REST API `GET /api/jobs/job_id` and the new job_submission snapshot are now the same.  This required a bit of a refactor, @edoakes could you please review this again?

**Review by edoakes — APPROVED** [review]

File: dashboard/modules/job/tests/test_job_manager.py
Comment (edoakes) [inline dashboard/modules/job/tests/test_job_manager.py:99]: nit: should this be renamed `job_manager.get_job_data` ?

Comment (edoakes) [reply dashboard/modules/job/tests/test_job_manager.py:99]: nvm, separate method, I see

**Pull request merged**
